### PR TITLE
Add some asserts about calldata encoded size.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libsolidity/codegen/CompilerUtils.cpp [hunk 1]
```diff
@@ -1001,6 +1001,7 @@ void CompilerUtils::storeStringData(bytesConstRef _data)
 unsigned CompilerUtils::loadFromMemoryHelper(Type const& _type, bool _fromCalldata, bool _padToWords)
 {
 	unsigned numBytes = _type.calldataEncodedSize(_padToWords);
+	solAssert(numBytes > 0, "");
```
Comment (axic) [inline libsolidity/codegen/CompilerUtils.cpp]: When can this be 0?

Comment (chriseth) [reply libsolidity/codegen/CompilerUtils.cpp]: In rare cases. Basically if it is zero, then it cannot be encoded. I think adding an assert right into the `calldataEncodedSize` function did not work in the past because it was called just to check.

The only cases where it can be zero are when it is a storage struct that contains a mapping or an array that contains such a thing.

Comment (axic) [reply libsolidity/codegen/CompilerUtils.cpp]: Can we have a test case which triggers this assert?

Comment (chriseth) [reply libsolidity/codegen/CompilerUtils.cpp]: If the assert is triggered, this means that the function was used incorrectly. We can create a unit test that uses the function incorrectly, but I don't really see the point...

File: libsolidity/codegen/CompilerUtils.cpp [hunk 2]
```diff
@@ -1054,6 +1055,7 @@ void CompilerUtils::rightShiftNumberOnStack(unsigned _bits, bool _isSigned)
 unsigned CompilerUtils::prepareMemoryStore(Type const& _type, bool _padToWords)
 {
 	unsigned numBytes = _type.calldataEncodedSize(_padToWords);
+	solAssert(numBytes > 0, "");
 	bool leftAligned = _type.category() == Type::Category::FixedBytes;
 	if (numBytes == 0)
```
Comment (axic) [inline libsolidity/codegen/CompilerUtils.cpp]: This is unreachable code.

Comment (chriseth) [reply libsolidity/codegen/CompilerUtils.cpp]: oh :)

File: libsolidity/codegen/CompilerUtils.cpp (hunk 1 above)
Comment (axic) [reply libsolidity/codegen/CompilerUtils.cpp]: Do we have proper checks in the typechecker which can avoid failing this assert?

File: libsolidity/codegen/CompilerUtils.cpp (hunk 2 above)
Comment (axic) [reply libsolidity/codegen/CompilerUtils.cpp]: Also in the other branch these is `number <= 32`. Could be merged into one proper assert up there.

File: libsolidity/codegen/CompilerUtils.cpp (hunk 1 above)
Comment (chriseth) [reply libsolidity/codegen/CompilerUtils.cpp]: This assert is the check. I'm not sure we are talking about the same thing...

**Comment by chriseth** [comment]
Updated.

File: libsolidity/codegen/CompilerUtils.cpp (hunk 2 above)
Comment (axic) [reply libsolidity/codegen/CompilerUtils.cpp]: This unreachable code is still there.

Comment (chriseth) [reply libsolidity/codegen/CompilerUtils.cpp]: Yes, but now it is reachable.

File: libsolidity/codegen/ArrayUtils.cpp [hunk 3]
```diff
@@ -291,8 +291,11 @@ void ArrayUtils::copyArrayToMemory(ArrayType const& _sourceType, bool _padToWord
 	CompilerUtils utils(m_context);
 	unsigned baseSize = 1;
 	if (!_sourceType.isByteArray())
+	{
 		// We always pad the elements, regardless of _padToWordBoundaries.
 		baseSize = _sourceType.baseType()->calldataEncodedSize();
+		solAssert(baseSize >= 0x20, "");
```
Comment (axic) [inline libsolidity/codegen/ArrayUtils.cpp:297]: Can you explain why is this needed?

Comment (chriseth) [reply libsolidity/codegen/ArrayUtils.cpp:297]: This might want to copy weird things like arrays of mappings. If baseSize is zero, then converting length into size and then looping over the size might not terminate.

**Review by axic — APPROVED** [review]
